### PR TITLE
Added transponder plugin

### PR DESCRIPTION
**Pull request by DoppleGangster**
Depends on https://github.com/mavlink/MAVSDK-Proto/pull/222

### CODE REVIEW

**Review by JonasVautherin — CHANGES_REQUESTED** [review]
Good start! Do you have a way to test that the conversions are correct? Like do you have a mavlink component that emits ADS-B messages?

**Comment by JonasVautherin** [comment]
One more thing: you need to fix the style, we have a script for that: `./tools/fix_style .` should do it :+1:

**Comment by DoppleGangster** [comment]
I thought I had already run the style script. I'll make sure to run it. I have an ADSB module ordered and it suppose to be delivered today.

**Comment by DoppleGangster** [comment]
Style should be correct now.

File: src/plugins/transponder/transponder_impl.cpp [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+void TransponderImpl::process_transponder(const mavlink_message_t& message)
+{
+    mavlink_adsb_vehicle_t local_adsb_vehicle;
+    mavlink_msg_adsb_vehicle_decode(&message, &local_adsb_vehicle);
+
+    Transponder::AdsbVehicle adsbVehicle;
+    adsbVehicle.icao_address = local_adsb_vehicle.ICAO_address;
+    adsbVehicle.latitude_deg = local_adsb_vehicle.lat;
+    adsbVehicle.longitude_deg = local_adsb_vehicle.lon;
```
[the remark this answers is not in the document]
Comment (JonasVautherin) [reply src/plugins/transponder/transponder_impl.cpp:69]: I'm a bit confused now... did the conversion get applied? I don't see them in the updated code :thinking:

File: src/plugins/transponder/include/plugins/transponder/transponder.h [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+        LighterAir, /**< @brief. */
+        Parachute, /**< @brief. */
+        UltraLight, /**< @brief. */
+        Unassigned2, /**< @brief. */
+        Uav, /**< @brief. */
+        Space, /**< @brief. */
+        Unassgined3, /**< @brief. */
+        EmergencySurface, /**< @brief. */
+        ServiceSurface, /**< @brief. */
+        PointObstacle, /**< @brief. */
```
[the remark this answers is not in the document]
Comment (DoppleGangster) [reply src/plugins/transponder/include/plugins/transponder/transponder.h:83]: I'm not familiar with which ones are actually used in the field so I wouldn't know which could be removed. I'll add comments and regenerate.

[the remark this answers is not in the document]
Comment (DoppleGangster) [reply src/plugins/transponder/include/plugins/transponder/transponder.h]: Should be updated now.

**Review by JonasVautherin — CHANGES_REQUESTED** [review]
Somehow all the conversions got lost in the process (i.e. all my suggestions). Did you revert them? :thinking:

**Comment by DoppleGangster** [comment]
I uploaded the files with the style changes. Not sure if that removes comments?

**Comment by DoppleGangster** [comment]
I think I know what happened. I'll go through and reimplement your recommendations tomorrow.
Sorry about that.

**Comment by JonasVautherin** [comment]
Also you get some unrelated CI issues, please rebase once https://github.com/mavlink/MAVSDK/pull/1392 is merged and that should solve them :+1:

**Comment by DoppleGangster** [comment]
@JonasVautherin: I've re-implemented your commits that were lost.

**Comment by JonasVautherin** [comment]
Starting to look good! Should we now wait until you can actually test it, and then move forward to merge it?

**Comment by DoppleGangster** [comment]
Sounds good to me. My expected receive date has gone to pending so most likely won't have the test results until sometime next week. Thanks for you help with this!

File: src/plugins/transponder/transponder_impl.cpp [hunk 3]
```diff
@@ -17,9 +17,22 @@ TransponderImpl::~TransponderImpl()
...
+        MAVLINK_MSG_ID_ADSB_VEHICLE,
+        std::bind(&TransponderImpl::process_transponder, this, _1),
+        this);
+}
 
-void TransponderImpl::deinit() {}
+void TransponderImpl::deinit() 
+{
+    _parent->unregister_statustext_handler(this);
+    _parent->unregister_param_changed_handler(this);
```
Comment (JonasVautherin) [inline src/plugins/transponder/transponder_impl.cpp:33]: Why are those needed? They are not registered in the `init`, are they? :thinking:

Comment (DoppleGangster) [reply src/plugins/transponder/transponder_impl.cpp]: Ahh yeah you're right. I'll remove them. I was using the telemetry_imp.cpp as a guide and might have gotten carried away with what I was bring over. I've received the sensor however my test code seems to have issues. My subscription that I create should output text at 1 Hz however I never get any output from the function.

Is there another tie in somewhere that I need to get mavsdk to register my subscription? 

[Created Example](https://github.com/DoppleGangster/MAVSDK/blob/main/examples/transponder/transponder.cpp)

**Comment by DoppleGangster** [comment]
@JonasVautherin: If I register for the subscription updates in the test program for the transponder API should I expect to get the values reported even if the transponder if not communicating? I get no errors when I set the update rate and subscription but it does not appear that the callback is ever executed.

**Comment by JonasVautherin** [comment]
> If I register for the subscription updates in the test program for the transponder API should I expect to get the values reported even if the transponder if not communicating?

When you subscribe, you will receive an update everytime the MAVLink message is sent. If your transponder does not send the MAVLink message, then nothing will happen. I'm expecting that the receiver is onboard the drone, receives ADS-B data and sends them as MAVLink messages on the sysid of the drone. Is that correct?

**Comment by DoppleGangster** [comment]
The receiver is onboard and I configured according to the [PX4 Air Traffic Setup](https://docs.px4.io/master/en/advanced_features/traffic_avoidance_adsb.html). I'm assuming this would be the same configuration to be able to access the MAVLINK messages for the ADS-B. The only difference in hardware is I'm using the NXP-FMU with a PingRX versus the mRO Pixhawk mentioned in the page.

**Comment by DoppleGangster** [comment]
@JonasVautherin : It does not appear that I have the ability to change the ADS-B module to use the drones sysid. Is there a way to have MAVSDK report messages of a certain type regardless of the sysid or do I need to find a way to grab the broadcast messages from the telemetry port that is communicating with offboard program?

**Comment by TSC21** [comment]
@DoppleGangster did you check if the firmware at runtime enters this loop: https://github.com/PX4/PX4-Autopilot/blob/4a7b2c490a2fea9d2f0a9a5805a1dbea3c0a3562/src/modules/mavlink/streams/ADSB_VEHICLE.hpp#L69. Have you tried accessing the nsh shell (can be done through the MAVLink shell in QGC) and type `listemer transponder_report`? If it states something like `never published` then something is wrong on your setup.
Before debugging MAVSDK, the first thing you need to do is making sure you actually have the transponder data being generated and propagated in PX4 and also being stream through MAVLink.

**Comment by DoppleGangster** [comment]
@TSC21 thanks for the comment. I confirmed that the data was never published. I've corrected the setup and now I am getting data from the transponder over MAVSDK.

@JonasVautherin and @TSC21 thank you both for your help with debugging this.

**Comment by DoppleGangster** [comment]
The plugin seems to be fully operation from my testing. Assuming no more changes are requested should be ready to merge with the proto fork update I would think?

**Comment by TSC21** [comment]
> The plugin seems to be fully operation from my testing. Assuming no more changes are requested should be ready to merge with the proto fork update I would think?

@DoppleGangster for both being ready for merge, you need to first squash the commits.

**Comment by JonasVautherin** [comment]
Good to read that it works!

I merged the MAVSDK-Proto PR. @DoppleGangster now you need to update the proto submodule such that it points to `main` on mavsdk-proto, and fix the style in this PR. Then we can merge it :+1:

**Review by JonasVautherin — CHANGES_REQUESTED** [review]
Good to read that it works!

I merged the MAVSDK-Proto PR. @DoppleGangster now you need to update the proto submodule such that it points to main on mavsdk-proto, and fix the style in this PR. Then we can merge it :+1: :rocket:

**Comment by DoppleGangster** [comment]
Proto module update is in and style updated. Thanks again!

**Review by JonasVautherin — APPROVED** [review]
